contracts/utils/OptionsPurchaseQueue.sol

```diff
@@ -0,0 +1,442 @@
+     * @param allocatedOptions Amount of options allocated
+     */
+    event OptionsAllocated(address indexed vault, uint256 allocatedOptions);
+
+    /**
+     * @notice Emitted when the vault sells options to the buyers
+     * @param vault The vault selling the options
+     * @param totalPremiums Total premiums earnt by the vault
+     * @param totalOptions Total options transferred to the buyers (allocatedOptions)
+     */
+    event OptionsSold(
+        address indexed vault,
+        uint256 totalPremiums,
+        uint256 totalOptions
+    );
+
+    /**
+     * @notice Emitted when a buyer is whitelisted for purchasing options
+     * @param buyer The whitelisted buyer
+     */
+    event BuyerWhitelisted(address indexed buyer);
+
+    /**
+     * @notice Emitted when a buyer is removed from the whitelist for purchasing options
+     * @param buyer The blacklisted buyer
+     */
+    event BuyerBlacklisted(address indexed buyer);
+
+    /**
+     * @notice Emitted when the ceiling price for a vault is updated
+     * @param vault The vault
+     * @param ceilingPrice The new ceiling price
+     */
+    event CeilingPriceUpdated(address indexed vault, uint256 ceilingPrice);
+
+    /************************************************
+     *  CONSTRUCTOR
+     ***********************************************/
+
+    /**
+     * @param _gnosisEasyAuction Gnosis Easy Auction contract
+     */
+    constructor(address _gnosisEasyAuction) {
+        gnosisEasyAuction = IGnosisAuction(_gnosisEasyAuction);
+    }
+
+    /************************************************
+     *  QUEUE OPERATIONS
+     ***********************************************/
+
+    /**
+     * @notice Create a request to purchase options from a vault at the auction settlement price
+     * @dev The buyer must be whitelisted to prevent the purchase queue from getting griefed (since sellToBuyers()
+     *  iterates through it). This function transfers the premiums for the options from the buyer at the ceiling
+     *  price (maximum price the buyer has to pay), however when the options are sold the buyer only pays the
+     *  auction settlement price and the leftover premiums are transferred back to the buyer. This function will
+     *  revert after the vault calls allocateOptions. New purchases can be made after the vault calls sellToBuyers().
+     *  The requests on the purchased queue are filled FIFO. Any unfilled/partially filled requests are refunded
+     *  their premiums, this can occur when the vault allocates less options than there are on the queue.
+     * @param vault The vault to purchase options from
+     * @param optionsAmount Amount of options requested
+     * @return premiums Amount of premiums transferred from the buyer
+     */
+    function requestPurchase(address vault, uint256 optionsAmount)
+        external
+        override
+        returns (uint256)
+    {
+        require(whitelistedBuyer[msg.sender], "Buyer not whitelisted");
```

Comment:
Probably better to set an array limit on `purchases[vault]` than whitelist the buyer so we don't have an unbounded for loop

Reply:
Switched to using a min purchase requirement for each purchase request